service/src/main/kotlin/fi/espoo/evaka/varda/VardaUpdateServiceV2.kt

```diff
@@ -0,0 +1,698 @@
+                alkamis_pvm = calculateVardaFeeDataStartDate(decision.validFrom, evakaServiceNeedInfoForVarda.asPeriod),
+                paattymis_pvm = minEndDate(evakaServiceNeedInfoForVarda.endDate, decision.validTo),
+                lahdejarjestelma = client.sourceSystem
+            )
+        )
+    } catch (e: Exception) {
+        throw Exception("VardaUpdate: cannot create voucher fee data ${decision.id} for service need ${evakaServiceNeedInfoForVarda.id}: varda client threw ${e.localizedMessage}")
+    }
+
+    if (res == null) throw Exception("VardaUpdate: cannot create voucher fee data ${decision.id} for service need ${evakaServiceNeedInfoForVarda.id}: create varda placement response is null")
+    return res.id
+}
+
+private fun vardaFeeBasisByPlacementType(type: PlacementType): String {
+    return if (type == PlacementType.DAYCARE_FIVE_YEAR_OLDS || type == PlacementType.DAYCARE_PART_TIME_FIVE_YEAR_OLDS)
+        FeeBasisCode.FIVE_YEAR_OLDS_DAYCARE.code
+    else
+        FeeBasisCode.DAYCARE.code
+}
+
+private fun calculateVardaFeeDataStartDate(fdStartDate: LocalDate, serviceNeedDates: DateRange): LocalDate {
+    return if (serviceNeedDates.includes(fdStartDate)) fdStartDate else serviceNeedDates.start
+}
+
+fun calculateEvakaVsVardaServiceNeedChangesByChild(db: Database.Connection, startingFrom: HelsinkiDateTime): Map<UUID, VardaChildCalculatedServiceNeedChanges> {
+    val evakaServiceNeedDeletionsByChild = calculateDeletedChildServiceNeeds(db)
+    val evakaServiceNeedChangesByChild = db.read { it.getEvakaServiceNeedChanges(startingFrom) }
+        .groupBy { it.evakaChildId }
+
+    val additionsAndChangesToVardaByChild = evakaServiceNeedChangesByChild.entries.associate { evakaServiceNeedChangesForChild ->
+        val vardaServiceNeedsForChild = db.read { it.getChildVardaServiceNeeds(evakaServiceNeedChangesForChild.key) }
+
+        evakaServiceNeedChangesForChild.key to VardaChildCalculatedServiceNeedChanges(
+            childId = evakaServiceNeedChangesForChild.key,
+            additions = calculateNewChildServiceNeeds(evakaServiceNeedChangesForChild.value, vardaServiceNeedsForChild),
+            updates = calculateUpdatedChildServiceNeeds(evakaServiceNeedChangesForChild.value, vardaServiceNeedsForChild),
+            deletes = evakaServiceNeedDeletionsByChild.getOrDefault(evakaServiceNeedChangesForChild.key, emptyList())
+        )
+    }
+
+    return additionsAndChangesToVardaByChild.plus(
+        evakaServiceNeedDeletionsByChild.filterNot {
+            additionsAndChangesToVardaByChild.containsKey(it.key)
+        }.entries.associate { childServiceNeedDeletions ->
+            childServiceNeedDeletions.key to
+                VardaChildCalculatedServiceNeedChanges(
+                    childId = childServiceNeedDeletions.key,
+                    additions = emptyList(),
+                    updates = emptyList(),
+                    deletes = childServiceNeedDeletions.value
+                )
+        }
+    )
```

Comment:
At least my poor brain finds it a bit difficult to understand what is happening here. Would it be too noisy to introduce one or two intermediary `val`'s here to make this more self-explanatory?